src/LeaseManager.cc

```diff
@@ -0,0 +1,307 @@
+ */
+void
+LeaseManager::LeaseCleaner::handleTimerEvent()
+{
+    bool stillCleaning = true;
+    while (stillCleaning) {
+        stillCleaning = leaseManager->cleanNextLease();
+    }
+    // Run once per lease term as some will likely have expired by then.
+    this->start(Cycles::rdtsc() + Cycles::fromNanoseconds(
+            LeaseCommon::LEASE_TERM_US * 1000));
+}
+
+/**
+ * Persist the next available leaseId to external storage=.
+ *
+ * \param lock
+ *      Ensures that caller has acquired mutex; not actually used here.
+ */
+void
+LeaseManager::allocateNextLease(Lock &lock)
+{
+    uint64_t nextLeaseId = maxAllocatedLeaseId + 1;
+    std::string leaseObjName = STORAGE_PREFIX + "/"
+            + format("%lu", nextLeaseId);
+    std::string str;    // TODO(cstlee): Is there a better way set an empty obj?
+    context->externalStorage->set(ExternalStorage::Hint::CREATE,
+                                  leaseObjName.c_str(),
+                                  str.c_str(), 0);
+    maxAllocatedLeaseId = nextLeaseId;
+}
+
+/**
+ * Expire the next lease whose term has elapsed. If the term of the next lease
+ * to be expired has not yet elapsed, this call has no effect.
+ *
+ * \return
+ *      Return true if a lease was able to be cleaned.  Returning false implies
+ *      there are no more leases that can be cleaned at this moment.
+ */
+bool
+LeaseManager::cleanNextLease()
+{
+    Lock lock(mutex);
+    ExpirationOrderSet::iterator it = expirationOrder.begin();
+    if (it != expirationOrder.end() && it->leaseTerm < clock.getTime()) {
+        uint64_t leaseId = it->leaseId;
+        std::string leaseObjName = STORAGE_PREFIX + "/"
+                + format("%lu", leaseId);
+        context->externalStorage->remove(leaseObjName.c_str());
+        expirationOrder.erase(it);
+        leaseMap.erase(leaseId);
+        return true;
+    }
+    return false;
+}
+
+/**
+ * Does most of the work for "renewLease".  Breaks up code for ease of testing.
+ *
+ * \param leaseId
+ *      See "renewLease".
+ * \param lock
+ *      Ensures that caller has acquired mutex; not actually used here.
+ * \return
+ *      See "renewLease".
+ */
+WireFormat::ClientLease
+LeaseManager::renewLeaseInternal(uint64_t leaseId, Lock &lock)
+{
+    WireFormat::ClientLease clientLease;
+
+    LeaseMap::iterator leaseEntry = leaseMap.find(leaseId);
+    if (leaseEntry != leaseMap.end()) {
+        // Simply renew the existing lease.
+        clientLease.leaseId = leaseId;
+        expirationOrder.erase({leaseEntry->second, leaseId});
+    } else {
+        // Must issue a new lease as the old one has expired.
+        clientLease.leaseId = ++lastIssuedLeaseId;
+
+        // The maxAllocatedLeaseId needs to always stay ahead of any issued
+        // leaseId.  If it is not, more leases need to be allocated.
+        // This should only ever execute once if at all.
+        while (maxAllocatedLeaseId <= lastIssuedLeaseId) {
+            allocateNextLease(lock);
```

Comment:
Why persist the lease "manually" rather than waiting for the pre-allocator to do it?